move GroupByFilename Dataset to DataPipe

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51709 move GroupByFilename Dataset to DataPipe**
* #51704 move RoutedDecoder Dataset to DataPipe

Move GroupByFilename Dataset to DataPipe

Differential Revision: [D26263585](https://our.internmc.facebook.com/intern/diff/D26263585)

Commit:
move GroupByFilename Dataset to DataPipe

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/iter/groupbyfilename.py:
```diff
@@ -0,0 +1,102 @@
+        assert self.group_size > 0
+        assert self.max_buffer_size >= 0
+
+        if self.group_size == 1:
+            for data in self.datapipe:
+                yield [data]
+
+        for data in self.datapipe:
+            key = self.group_key_fn(data)
+            res = self.stream_buffer.get(key, []) + [data]
+            if len(res) == self.group_size:
+                yield self.sort_data_fn(res)
+                del self.stream_buffer[key]
+                self.curr_buffer_size = self.curr_buffer_size - self.group_size + 1
+            else:
+                if self.curr_buffer_size == self.max_buffer_size:
+                    raise OverflowError(
+                        "stream_buffer is overflow, please adjust the order of data "
+                        "in the input datapipe or increase the buffer size!")
+                self.stream_buffer[key] = res
+                self.curr_buffer_size = self.curr_buffer_size + 1
```
I am not sure if it will cause some problem. If someone implements some hack to raise StopIteration in the middle of iteration of `datapipe`, this generator will go from the part of `group_size == 1` to the second part. So I guess it's safer to use `if ... else` here.

ejguan commented on torch/utils/data/datapipes/iter/groupbyfilename.py:
```diff
@@ -0,0 +1,102 @@
+
+
+    def __iter__(self) -> Iterator[list]:
+        assert self.group_size > 0
+        assert self.max_buffer_size >= 0
```
BTW, `max_buffer_size` should be larger or equal to group_size.

glaringlee commented on torch/utils/data/datapipes/iter/groupbyfilename.py:
```diff
@@ -0,0 +1,102 @@
+    return sorted(datalist, key=functools.cmp_to_key(cmp_fn))
+
+
+class GroupByFilenameIterDataPipe(IterDataPipe):
+    r""" :class:`GroupByFilenameIterDataPipe`.
+
+    Iterable datapipe to group binary streams from input iterables by pathname without extension,
+    yield a list with `group_size` items in it, each item in the list is a tuple of
+    pathname and binary stream
```
Let me just change this to `GroupByKeyIterDataPipe`, I don't think we need a base class.  What will be used as a key is totally decided by the `group_key_fn`. I will change the comments in the code as well.

Commit:
Update on "move GroupByFilename Dataset to DataPipe"


Move GroupByFilename Dataset to DataPipe

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/iter/groupbykey.py:
```diff
@@ -0,0 +1,107 @@
+        else:
+            for data in self.datapipe:
+                key = self.group_key_fn(data)
+                res = self.stream_buffer.get(key, []) + [data]
```
We can avoid copy here.
```python
if key not in self.stream_buffer:
    self.stream_buffer[key] = []
res = self.stream_buffer[key]  # share same storage
res.append(data) 
```

ejguan commented on torch/utils/data/datapipes/iter/groupbykey.py:
```diff
@@ -0,0 +1,107 @@
+                        raise OverflowError(
+                            "stream_buffer is overflow, please adjust the order of data "
+                            "in the input datapipe or increase the buffer size!")
+                    self.stream_buffer[key] = res
```
So it's not needed here.

ejguan reviewed (approved):
LGTM with one nit optimization.

Commit:
Update on "move GroupByFilename Dataset to DataPipe"


Move GroupByFilename Dataset to DataPipe

Differential Revision: [D26263585](https://our.internmc.facebook.com/intern/diff/D26263585)

[ghstack-poisoned]

Commit:
Update on "move GroupByFilename Dataset to DataPipe"


Move GroupByFilename Dataset to DataPipe

Differential Revision: [D26263585](https://our.internmc.facebook.com/intern/diff/D26263585)

[ghstack-poisoned]
